File: 198-house-robber.java
class Solution {
    /*public int rob(int[] nums) {
        if(nums == null || nums.length == 0) { return 0;}
        if(nums.length < 2) { return nums[0]; }
        
        int[] dp = new int[nums.length];
        dp[0] = nums[0];
        dp[1] = Math.max(nums[0], nums[1]);
        
        for(int house = 2; house < nums.length; house++) {
            dp[house] = Math.max(dp[house - 1], dp[house - 2] + nums[house]);
        }
        
        return dp[nums.length - 1];        
    }*/
    
    public int rob(int[] nums) {
        if(nums == null || nums.length == 0) { return 0;}
        if(nums.length < 2) { return nums[0]; }
        
        int moneyFromPrevious1 = Math.max(nums[0], nums[1]);
        int moneyFromPrevious2 = nums[0];   
        
        int money = 0;
        for(int house = 2; house < nums.length; house++) {
            money = Math.max(moneyFromPrevious1, moneyFromPrevious2 + nums[house]);
            moneyFromPrevious2 = moneyFromPrevious1;
            moneyFromPrevious1 = money;           
        }
        
        return moneyFromPrevious1;        
    }    
}

File: 198_House_Robber.py

This problem is about making the best choice at each house.
For every house, you decide between two options:

Rob this house → but then you must skip the previous one

Skip this house → and keep whatever the best total was so far

So for each house, you compare:

the best total if you don’t take it

the best total if you do take it, which includes the house’s money plus the best total from two houses back

By walking through the street once and always choosing the better of these two options, you get the maximum money without ever robbing two adjacent houses.

class Solution:
    def rob(self, nums: List[int]) -> int:
        rob_prev = 0      # max up to previous house
        rob_prev2 = 0     # max up to two houses ago

        for money in nums:
            rob_current = max(rob_prev, rob_prev2 + money)
            rob_prev2 = rob_prev
            rob_prev = rob_current

        return rob_prev
